k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapFolder.java

```diff
@@ -1191,12 +1193,32 @@ public String getUidFromMessageId(Message message) throws MessagingException {
     }
 
     @Override
-    public void expunge() throws MessagingException {
+    public void expunge(List<Long> knownDeletedUids) throws MessagingException {
+        if (knownDeletedUids == null || knownDeletedUids.isEmpty()) {
+            return;
+        }
         open(OPEN_MODE_RW);
         checkOpen();
 
         try {
-            executeSimpleCommand("EXPUNGE");
+            if (connection.isUidPlusCapable()) {
+                UidExpungeCommand uidExpungeCommand = new UidExpungeCommand.Builder()
+                        .idSet(knownDeletedUids)
+                        .build();
+
+                uidExpungeCommand.execute(connection, this);
+            } else {
+                UidSearchCommand searchCommand = new UidSearchCommand.Builder()
```

Comment:
I have moved the UidSearchCommand and it's execution into a new method `getUnknownDeletedUids()` and have added a comment regarding the RFC at the top of the try block